consensus/p2p/validator/proposal_stream_demux.go

```diff
@@ -0,0 +1,239 @@
+		currentHeight:    currentHeight,
+	}
+}
+
+func (t *proposalStreamDemux) Loop(ctx context.Context, topic *pubsub.Topic) {
+	defer close(t.outputs)
+
+	messages := make(chan *pubsub.Message, t.bufferSizeConfig.ProposalDemux)
+	defer close(messages)
+
+	// Create a new pool for the current height
+	t.createCtxPool(ctx)
+	defer t.stop()
+
+	wg := conc.NewWaitGroup()
+
+	// Process the stream messages and the commits sequentially
+	wg.Go(func() {
+		for {
+			select {
+			case <-ctx.Done():
+				return
+			case message := <-messages:
+				if err := t.processStreamMessage(ctx, message); err != nil {
+					t.log.Errorw("error processing stream message", "error", err)
+				}
+			case height := <-t.commitNotifier:
+				// Stop the current height pool and create a new one
+				t.stop()
+				t.createCtxPool(ctx)
+
+				// Process the commit
+				if err := t.processCommit(height); err != nil {
+					t.log.Errorw("error processing commit", "error", err)
+				}
+			}
+		}
+	})
+
+	// Subscribe to the topic and forward messages to a channel
+	wg.Go(func() {
+		onMessage := func(ctx context.Context, message *pubsub.Message) {
+			select {
+			case <-ctx.Done():
+				return
+			case messages <- message:
+			}
+		}
+		topicSubscription := buffered.NewTopicSubscription(t.log, t.bufferSizeConfig.ProposalDemux, onMessage)
+		topicSubscription.Loop(ctx, topic)
+	})
+
+	wg.Wait()
+}
+
+func (t *proposalStreamDemux) Listen() <-chan starknet.Proposal {
+	return t.outputs
+}
+
+// stop stops the current height pool and deletes the streams for the current height
+func (t *proposalStreamDemux) stop() {
+	// First cancel all running stream of the current height
+	t.currentHeightCancel()
+
+	// Wait for all the streams to finish. Error of each stream is logged, so we ignore the combined error.
+	_ = t.currentHeightCtxPool.Wait()
+
+	// Close the input channels and delete the stream objects
+	for _, streamID := range t.streamHeights[t.currentHeight] {
+		t.streams[streamID].close()
+		delete(t.streams, streamID)
+	}
+
+	// Finally delete the list of streams for the old height
+	delete(t.streamHeights, t.currentHeight)
+}
+
+// createCtxPool creates a new pool for the current height
+func (t *proposalStreamDemux) createCtxPool(ctx context.Context) {
+	ctx, t.currentHeightCancel = context.WithCancel(ctx)
+	t.currentHeightCtxPool = pool.New().WithContext(ctx)
+}
+
+// processCommit processes the commit for the current height
+func (t *proposalStreamDemux) processCommit(height types.Height) error {
+	if height != t.currentHeight {
+		return fmt.Errorf("expected height %d, got %d", t.currentHeight+1, height)
+	}
+
+	// Start the all the streams for the new height
+	t.currentHeight++
+	for _, streamID := range t.streamHeights[t.currentHeight] {
+		stream := t.streams[streamID]
+		t.currentHeightCtxPool.Go(func(ctx context.Context) error {
+			stream.loop(ctx)
+			return nil
+		})
+	}
+
+	return nil
+}
+
```

Comment:
nit: feel free to ignore
```suggestion
	// If the msg is for this height, then start a go-routine
	// that will process this particular proposal in a loop
	if message.SequenceNumber == 0 {
		return t.onFirstMessage(ctx, streamID, &message)
	}

	// Forward the proposal onto the go-routine started above
	return t.onSubsequentMessage(ctx, streamID, &message)
}
```